vmm/task/src/youki.rs

```diff
@@ -0,0 +1,646 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+use std::{
+    convert::TryFrom,
+    os::fd::{FromRawFd, IntoRawFd, OwnedFd, RawFd},
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+use async_trait::async_trait;
+use containerd_shim::{
+    api::{CreateTaskRequest, ExecProcessRequest, Options, Status},
+    asynchronous::{
+        console::ConsoleSocket,
+        container::{ContainerFactory, ContainerTemplate, ProcessFactory},
+        processes::{ProcessLifecycle, ProcessTemplate},
+    },
+    io::Stdio,
+    io_error, other, other_error,
+    protos::{
+        api::ProcessInfo,
+        cgroups::metrics::Metrics,
+        protobuf::{CodedInputStream, Message},
+    },
+    util::read_spec,
+    Error, ExitSignal, Result,
+};
+use libcontainer::{
+    container::{builder::ContainerBuilder, Container},
+    error::LibcontainerError,
+    signal::Signal,
+    syscall::syscall::SyscallType,
+};
+use log::{debug, warn};
+use nix::{
+    sys::signal::kill,
+    unistd::{Gid, Pid, Uid},
+};
+use oci_spec::runtime::{LinuxResources, Process, Spec};
+use runc::io::{IOOption, Io, NullIo};
+use tokio::{
+    io::{AsyncRead, AsyncWrite},
+    process::Command,
+    sync::Mutex,
+    task::spawn_blocking,
+};
+use vmm_common::{storage::Storage, KUASAR_STATE_DIR};
+
+use crate::{
+    device::rescan_pci_bus,
+    io::{convert_stdio, copy_io_or_console, ProcessIO},
+    sandbox::SandboxResources,
+    util::{read_io, read_storages},
+};
+
+pub type ExecProcess = ProcessTemplate<YoukiExecLifecycle>;
+pub type InitProcess = ProcessTemplate<YoukiInitLifecycle>;
+
+pub type YoukiContainer = ContainerTemplate<InitProcess, ExecProcess, YoukiExecFactory>;
+
+pub const YOUKI_DIR: &str = "/run/kuasar/youki";
+const STORAGE_ANNOTATION: &str = "io.kuasar.storages";
```

Comment:
`STORAGE_ANNOTATION` should not be a private variable of `youki.rs`, I think we have a public on in `vmm/common`

Reply:
done